### PR TITLE
torch check the division by zero in batch_norm_update_stats

### LINKED ISSUE #120803: Floating Point Exception on Empty Input Tensor with `torch.batch_norm_update_stats`
## Describe the bug
When attempting to update batch normalization statistics using `torch.batch_norm_update_stats` with an input tensor of shape `[0,0]`, a floating point exception (core dumped) error occurs. 

## Error log
```bash
Floating point exception (core dumped)
```

## Minified repro
```python
import torch
inp = torch.rand([0,0])
running_mean = torch.rand([1])
running_var = torch.rand([1])
torch.batch_norm_update_stats(input=inp, momentum=0.0, running_mean=running_mean, running_var = running_var)
```

### Versions

PyTorch version: 2.3.0.dev20240220+cu118
Is debug build: False
CUDA used to build PyTorch: 11.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.35

Python version: 3.8.17 (default, Jul 5 2023, 21:04:15) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.5.0-15-generic-x86_64-with-glibc2.17
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture: x86_64
CPU op-mode(s): 32-bit, 64-bit
Address sizes: 48 bits physical, 48 bits virtual
Byte Order: Little Endian
CPU(s): 256
On-line CPU(s) list: 0-255
Vendor ID: AuthenticAMD
Model name: AMD EPYC 7763 64-Core Processor
CPU family: 25
Model: 1
Thread(s) per core: 2
Core(s) per socket: 64
Socket(s): 2
Stepping: 1
Frequency boost: enabled
CPU max MHz: 3529.0520
CPU min MHz: 1500.0000
BogoMIPS: 4890.76
Flags: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ht syscall nx mmxext fxsr_opt pdpe1gb rdtscp lm constant_tsc rep_good nopl nonstop_tsc cpuid extd_apicid aperfmperf rapl pni pclmulqdq monitor ssse3 fma cx16 pcid sse4_1 sse4_2 x2apic movbe popcnt aes xsave avx f16c rdrand lahf_lm cmp_legacy svm extapic cr8_legacy abm sse4a misalignsse 3dnowprefetch osvw ibs skinit wdt tce topoext perfctr_core perfctr_nb bpext perfctr_llc mwaitx cpb cat_l3 cdp_l3 invpcid_single hw_pstate ssbd mba ibrs ibpb stibp vmmcall fsgsbase bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a rdseed adx smap clflushopt clwb sha_ni xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local clzero irperf xsaveerptr rdpru wbnoinvd amd_ppin brs arat npt lbrv svm_lock nrip_save tsc_scale vmcb_clean flushbyasid decodeassists pausefilter pfthreshold v_vmsave_vmload vgif v_spec_ctrl umip pku ospke vaes vpclmulqdq rdpid overflow_recov succor smca fsrm
Virtualization: AMD-V
L1d cache: 4 MiB (128 instances)
L1i cache: 4 MiB (128 instances)
L2 cache: 64 MiB (128 instances)
L3 cache: 512 MiB (16 instances)
NUMA node(s): 2
NUMA node0 CPU(s): 0-63,128-191
NUMA node1 CPU(s): 64-127,192-255
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit: Not affected
Vulnerability L1tf: Not affected
Vulnerability Mds: Not affected
Vulnerability Meltdown: Not affected
Vulnerability Mmio stale data: Not affected
Vulnerability Retbleed: Not affected
Vulnerability Spec rstack overflow: Mitigation; safe RET
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1: Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2: Mitigation; Retpolines, IBPB conditional, IBRS_FW, STIBP always-on, RSB filling, PBRSB-eIBRS Not affected
Vulnerability Srbds: Not affected
Vulnerability Tsx async abort: Not affected

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240220+cu118
[pip3] torchaudio==2.2.0.dev20240220+cu118
[pip3] torchvision==0.18.0.dev20240220+cu118
[conda] numpy 1.23.5 pypi_0 pypi
[conda] pytorch-triton 3.0.0+901819d2b6 pypi_0 pypi
[conda] torch 2.3.0.dev20240220+cu118 pypi_0 pypi
[conda] torchaudio 2.2.0.dev20240220+cu118 pypi_0 pypi
[conda] torchvision 0.18.0.dev20240220+cu118 pypi_0 pypi

cc @malfet

### PR DESCRIPTION
**Pull request by lancerts**
Fixes #120803

### CODE REVIEW

File: aten/src/ATen/native/Normalization.cpp [hunk 1]
```diff
@@ -201,6 +201,8 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
   int64_t n_input = input.size(1);
   int64_t n = input.numel() / n_input;
 
+  TORCH_CHECK(n_input > 0 && input.numel() > 0, "both input.size(1) and input.numel() need to be greater than 0, got input.size(1) = ", n_input, ", input.numel() = ", input.numel());
```
Comment (mingfeima) [inline aten/src/ATen/native/Normalization.cpp:204]: Move the check before line 202 ?

**Review by mingfeima — CHANGES_REQUESTED** [review]
also need to update the test cases.

**Comment by lancerts** [comment]
> also need to update the test cases.

sure thing. Didnt find exsisting test cases for `batch_norm_update_stats`. Just to confirm, should i add the test  cases in `test/test_nn.py`?

**Comment by lancerts** [comment]
> also need to update the test cases.

@mingfeima test cases added, ty.

File: aten/src/ATen/native/Normalization.cpp [hunk 2]
```diff
@@ -199,6 +199,7 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
   using accscalar_t = at::acc_type<scalar_t, false>;
 
   int64_t n_input = input.size(1);
+  TORCH_CHECK(n_input > 0 && input.numel() > 0, "both input.size(1) and input.numel() need to be greater than 0, got input.size(1) = ", n_input, ", input.numel() = ", input.numel());
```
Comment (CaoE) [inline aten/src/ATen/native/Normalization.cpp:202]: Is there any case that: input.numel() >0 but n_input ==0 ?

Comment (lancerts) [reply aten/src/ATen/native/Normalization.cpp:202]: Yes, if `input = torch.rand(1, 0)`.

Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:202]: If `input = torch.rand(1, 0)` the results are:
n_input : 0
input.numel() : 0

Comment (lancerts) [reply aten/src/ATen/native/Normalization.cpp:202]: My bad, typo, should be 
```
input = torch.rand(0, 1)


print("input.size(0): ", input.size(0))
print("input.size(1): ", input.size(1))
print("input.numel(): ", input.numel())
```
outputs
```
input.size(0):  0
input.size(1):  1
input.numel():  0
```

Comment (lancerts) [reply aten/src/ATen/native/Normalization.cpp:202]: @CaoE misread the question, sry. You are right, i dont think it is possible ` input.numel() >0 but n_input ==0`. input.numel() >0 implies all dimension has size>0.

Comment (lancerts) [reply aten/src/ATen/native/Normalization.cpp:202]: Do you suggest checking only `input.numel() >0`? Both input.numel() and n_input are used as denominator so it is safer to check both variables. Understand it might be redundent. 
Not sure if there are corner cases  with ` input.numel() >0 but n_input ==0` (i couldnt think one).

Comment (CaoE) [reply aten/src/ATen/native/Normalization.cpp:202]: There is also `int64_t image_size = input.numel() / input.size(0) / input.size(1);` in https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/cpu/batch_norm_kernel.cpp#L1354.
I think check `input.numel() >0` is enough or you need also check `input.size(0)`.

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 3]
```diff
@@ -722,6 +722,16 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda(
...
+
+  TORCH_CHECK(
+      self.size(0) > 0 && n_input > 0 && self.numel() > 0,
+      "input.size(0), input.size(1) and input.numel() need to be greater than 0, got input.size(0) = ",
+      self.size(0),
+      ", input.size(1) = ",
+      n_input,
+      ", input.numel() = ",
+      self.numel()
+  );
```
Comment (CaoE) [inline aten/src/ATen/native/cuda/Normalization.cu:734]: Will there be the same issue as https://github.com/pytorch/pytorch/issues/120803 on CUDA ?

Comment (lancerts) [reply aten/src/ATen/native/cuda/Normalization.cu:734]:
```
import torch
input = torch.rand(0, 1)
input.to("cuda")
running_mean = torch.rand(1)
running_var = torch.rand(1)

running_mean.to("cuda")
running_var.to("cuda")
torch.batch_norm_update_stats(input=input, momentum=0.0, running_mean=running_mean, running_var = running_var)
```

gives
```
[1]    21652 floating point exception  /home/lancerts/miniconda3/envs/wsl-dev/bin/python
```

Comment (CaoE) [reply aten/src/ATen/native/cuda/Normalization.cu:734]: Got it. Thanks !

**Review by CaoE — APPROVED** [review]

**Comment by lancerts** [comment]
@malfet for final checks, thanks for the help

**Comment by lancerts** [comment]
@malfet for final checks, thanks for the help

File: aten/src/ATen/native/Normalization.cpp [hunk 4]
```diff
@@ -199,6 +199,15 @@ std::tuple<Tensor,Tensor> batch_norm_cpu_update_stats_template(
...
   int64_t n_input = input.size(1);
+  TORCH_CHECK(
+      input.size(0) > 0 && n_input > 0 && input.numel() > 0,
+      "input.size(0), input.size(1) and input.numel() need to be greater than 0, got input.size(0) = ",
+      input.size(0),
+      ", input.size(1) = ",
+      n_input,
+      ", input.numel() = ",
+      input.numel()
+  );
```
Comment (malfet) [inline aten/src/ATen/native/Normalization.cpp:210]: if size of any of the dimensions is 0, therefore checking only for `input.numel()` is sufficient. And along the same venue, why not just print a tensor shape?
```suggestion
  TORCH_CHECK(input.numel() != 0, "input must be a non-empty tensor, but got  ", input.sizes());
```

File: aten/src/ATen/native/cuda/Normalization.cu (hunk 3 above)
Comment (malfet) [inline aten/src/ATen/native/cuda/Normalization.cu:734]: Same as the above, use Occam's razor and device a simplest check (and perhaps error message)

Comment (lancerts) [reply aten/src/ATen/native/cuda/Normalization.cu]: Modified.

File: aten/src/ATen/native/Normalization.cpp (hunk 4 above)
Comment (lancerts) [reply aten/src/ATen/native/Normalization.cpp]: @malfet Thanks for the confirmation that numel check is enough, was afraid missing some corner cases. Modified to  `TORCH_CHECK(input.numel() != 0, "input tensor must have at least one element, but got input_sizes = ", input.sizes());`

Did not use non-empty tensor since people may conflate this with torch.empty...
For example, input with torch.empty([1,1]) is OK. (It is actually not an "empty" tensor but people may not know due to the name).

**Review by malfet — APPROVED** [review]

File: aten/src/ATen/native/cuda/Normalization.cu [hunk 5]
```diff
@@ -722,6 +722,8 @@ std::tuple<Tensor, Tensor> batch_norm_update_stats_cuda(
   c10::MaybeOwned<Tensor> running_var = at::borrow_from_optional_tensor(running_var_opt);
 
   const int64_t n_input = self.size(1);
+
+  TORCH_CHECK(self.numel() != 0, "input tensor must have at least one element, but got input_sizes = ", input.sizes());
```
Comment (malfet) [inline aten/src/ATen/native/cuda/Normalization.cu:726]:
```suggestion
  TORCH_CHECK(self.numel() != 0, "input tensor must have at least one element, but got input_sizes = ", self.sizes());
```
